### PR TITLE
[processor/transform] Add ability to truncate attributes

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:**
This PR adds to the transform processor the ability to truncate string values in maps.  The intended goal being the ability to truncate string values on attributes and resource.attributes.

**Link to tracking Issue:**
#9366
This is one of two PR I plan to submit for this issue. 

**Testing:**
Added unit tests and tested locally running the collector and tracegen

**Documentation:**
Updated README

### CODE REVIEW

File: processor/transformprocessor/README.md [hunk 1]
```diff
@@ -43,6 +45,8 @@ processors:
         - set(status.code, 1) where attributes["http.path"] == "/health"
         - keep_keys(resource.attributes, "service.name", "service.namespace", "cloud.region")
         - set(name, attributes["http.route"])
+        - truncate(attributes, 4096)
+        - truncate(resource.attributes, 4096)
```
Comment (bogdandrutu) [inline processor/transformprocessor/README.md:49]: "truncateAll"? Since I expect to have a truncate for a specific key?

**Review by bogdandrutu — COMMENTED** [review]
/cc @anuraaga

File: processor/transformprocessor/README.md (hunk 1 above)
Comment (TylerHelmuth) [reply processor/transformprocessor/README.md:54]: Good thinking.  I will update the name.

File: processor/transformprocessor/internal/common/functions.go [hunk 2]
```diff
@@ -70,6 +71,32 @@ func keepKeys(target GetSetter, keys []string) ExprFunc {
 	}
 }
 
+func truncateAll(target GetSetter, limit int64) ExprFunc {
```
Comment (mx-psi) [inline processor/transformprocessor/internal/common/functions.go:74]: Limit should be an unsigned integer (or at least passing negative integers should make the transform processor fail validation)

Comment (anuraaga) [reply processor/transformprocessor/internal/common/functions.go:74]: The parser currently always returns an `int64` because it doesn't know semantics yet, which is what's passed here. So indeed, validation here would be good - currently the factories return `ExprFunc` but it seems they should actually be `ExprFunc, error`

Comment (mx-psi) [reply processor/transformprocessor/internal/common/functions.go:74]: If this is not done anywhere else, it's fine for me to leave this for a different PR

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/common/functions.go:74]: I'd not in place yet, but I see how it could be done.  I will open a new issue to add that functionality to the framework.  In the meantime I added a check in the function itself so that limits < 0 do nothing.  I also added a new test.

**Review by mx-psi — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]

**Review by pmm-sumo — APPROVED** [review]

File: processor/transformprocessor/README.md [hunk 3]
```diff
@@ -22,6 +22,8 @@ it references an unset map value, there will be no action.
 - `keep_keys(target, string...)` - `target` is a path expression to a map type field. The map will be mutated to only contain
 the fields specified by the list of strings. e.g., `keep_keys(attributes, "http.method")`, `keep_keys(attributes, "http.method", "http.route")`
 
+- `truncateAll(target, limit)` - `target` is a path expression to a map type field. `limit` is an integer.  The map will be mutated such that all string values are truncated to the limit. e.g., `truncate(attributes, 100)` will truncate all string values in `attributes` such that all string values have less than or equal to 100 characters.  Non-string values are ignored.
```
Comment (codeboten) [inline processor/transformprocessor/README.md:25]: I'd suggest this be renamed to `truncate_all` or alternatively have `keep_keys` renamed to `keepKeys` for consistency

File: processor/transformprocessor/internal/common/functions.go [hunk 4]
```diff
@@ -117,6 +148,11 @@ func NewFunctionCall(inv Invocation, functions map[string]interface{}, pathParse
 				}
 				args = append(args, reflect.ValueOf(arg))
 				continue
+			case "int64":
+				if argDef.Int == nil {
+					return nil, fmt.Errorf("invalid argument at position %v, must be an int", i)
+				}
+				args = append(args, reflect.ValueOf(*argDef.Int))
```
Comment (codeboten) [inline processor/transformprocessor/internal/common/functions.go:155]: this change seem unrelated to the truncate functionality

**Review by codeboten — COMMENTED** [review]
Thanks for the improvement @TylerHelmuth, just a couple of questions

File: processor/transformprocessor/README.md (hunk 3 above)
Comment (TylerHelmuth) [reply processor/transformprocessor/README.md:25]: You're right, should be truncate_all

File: processor/transformprocessor/internal/common/functions.go (hunk 4 above)
Comment (TylerHelmuth) [reply processor/transformprocessor/internal/common/functions.go:155]: This addition allows the new  truncate_all function to be called with a second parameter that is an int.  Without this, the `args` variable won't contain the second parameter that is an int64.

Comment (codeboten) [reply processor/transformprocessor/internal/common/functions.go:155]: My bad, i was confused by the spacing in the github diff 👍

**Review by codeboten — APPROVED** [review]
